### PR TITLE
Simplify `CompactionOutputs::AddRangeDels()`

### PR DESCRIPTION
**Pull request by ajkr**
- The `{lower,upper}_bound` local variables in `AddRangeDels()` now represent internal keys. This makes `{lower,upper}_bound` directly usable for clamping `tombstone_{start,end}`
  - Tied to this change, adjacent subcompactions with a straddling range tombstone can now have the file preceding the boundary end at `kMaxSequenceNumber`, and the file succeeding the boundary begin at `kMaxSequenceNumber-1`. This is needed so that `{lower,upper}_bound` internal keys can be fully determined before seeing the tombstones. This eliminates special logic for `lower_bound_from_sub_compact`
- The scope of a range tombstone (per `FragmentedRangeTombstoneIterator::parsed_start_key()`) now starts at the tombstone seqno rather than `kMaxSequenceNumber`. This ensures a file's `range_tombstone_lower_bound_` does not overlap the preceding file's internal key range, which eliminates the special logic for `lower_bound_from_range_tombstone` 
- `TruncatedRangeDelMergingIter` now returns range tombstones whose scope overlaps an inclusive-inclusive internal key range. This helps `CompactionOutputs::AddRangeDels()` to query range tombstones in the output file's scope. But it does not fully solve the problem as the fragmenting iterator returned by `CompactionRangeDelAggregator::NewIterator()`  still may return range tombstones outside the desired scope; there is a TODO for pushing this logic down

Test Plan: unit test taken from @cbi42's #11067 and comments; correctness tests stressing file cutting such as the following command:

```
$ TEST_TMPDIR=/dev/shm python3 tools/db_crashtest.py blackbox --simple --max_key=100 --interval=10 --write_buffer_size=262144 --target_file_size_base=256 --max_bytes_for_level_base=262144 --block_size=128 --value_size_mult=33 --subcompactions=10
```

### CODE REVIEW

File: db/compaction/compaction_outputs.cc [hunk 1]
```diff
@@ -413,93 +413,93 @@ Status CompactionOutputs::AddRangeDels(
...
+    // `next_table_min_key`. The current file needs to extend range tombstone
+    // coverage through its own keys (through `meta.largest`) and through user
+    // keys preceding `next_table_min_key`'s user key.
+    ParsedInternalKey next_table_min_key_parsed;
+    ParseInternalKey(next_table_min_key, &next_table_min_key_parsed,
+                     false /* log_err_key */)
+        .PermitUncheckedError();
+    assert(next_table_min_key_parsed.sequence < kMaxSequenceNumber);
+
+    if (meta.largest.size() > 0 &&
```
[the remark this answers is not in the document]
Comment (cbi42) [reply db/compaction/compaction_outputs.cc:480]: unit tests as discussed offline: https://github.com/facebook/rocksdb/blob/cfaee55a8f4451f5f9cbc203e8b7aae3cbcbcd93/db/db_range_del_test.cc#L3180-L3357

**Review by cbi42 — APPROVED** [review]
LGTM once the case where `largest < smallest` is fixed. Thanks for the effort to simplify this complicated function! This should fix some bugs that I introduced in #10802 (essentially I did not properly take care of the `largest < smallest` case). This also helps a lot with the further simplifications in `TruncatedRangeDelIterator` and `MergingIterator` that I plan to do next.

File: db/compaction/compaction_outputs.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ajkr) [reply db/compaction/compaction_outputs.cc:488]: Done

**Review by ajkr — COMMENTED** [review]
Thanks for the review (and for writing all the tests)!

**Comment by ajkr** [comment]
>  I will try to merge only the tests and some comments after # 11078 is merged

Just was this on #11067. Will leave the tests for that PR.

BTW, even with this fix I saw TestIterate() trip an assertion in Prev():

```
db_stress: table/merging_iterator.cc:188: void rocksdb::MergingIterator::InsertRangeTombstoneToMaxHeap(size_t, bool, bool): Assertion `active_.count(level) == 0' failed.
```

The key it fails on (0x00000000000000090000000000000018) is a boundary key for several files:

```
--- level 1 --- version# 1 ---
...
 144354:1494[4689380 .. 4689768]['00000000000000090000000000000016787878787878' seq:4689747, type:1 .. '00000000000000090000000000000018' seq:72057594037927935, type:15]
 144356:1569[4689276 .. 4689741]['00000000000000090000000000000018' seq:4689613, type:1 .. '0000000000000009000000000000001A7878' seq:72057594037927935, type:15]
...
--- level 2 --- version# 1 ---
...
 144337:2739[4660048 .. 4687387]['00000000000000090000000000000016787878787878' seq:4660991, type:1 .. '00000000000000090000000000000018' seq:72057594037927935, type:15]
 144338:2397[4646458 .. 4687393]['00000000000000090000000000000018' seq:4687393, type:1 .. '0000000000000009000000000000001978' seq:4676883, type:1]
...
--- level 5 --- version# 1 ---
...
 143722:2251[0 .. 4649669]['00000000000000090000000000000013' seq:72057594037927934, type:15 .. '00000000000000090000000000000018' seq:72057594037927935, type:15]
 143728:1325[0 .. 0]['00000000000000090000000000000018' seq:0, type:1 .. '0000000000000009000000000000001978' seq:0, type:1]
...
```

**Comment by ajkr** [comment]
I think @cbi42 will incorporate this in a future PR

image

**Pull request closed without merging**